webapp/channels/src/components/suggestion/provider.tsx

```diff
@@ -1,20 +1,33 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 
-export type ProviderResult = {
+import {RequireOnlyOne} from '@mattermost/types/utilities';
+
+export type ProviderResult<Item> = {
     matchedPretext: string;
     terms: string[];
-    items: Array<Record<string, any>>;
-    component?: React.ReactNode;
-}
+    items: Array<Item | Loading>;
```

Comment:
This one took me a while to work out, but there's some cases where we return an object with `{loading: true}` to signify that more results are loading. The `SuggestionList` itself handles that case so the `*Suggestion` components only need to handle their `item` prop being an `Item`